### PR TITLE
parallel purger

### PR DESCRIPTION
**Pull request by jtaczanowski**

### CODE REVIEW

File: vaas-app/src/vaas/purger/purger.py [hunk 1]
```diff
@@ -12,23 +16,34 @@ def log_and_return_data(self, responses_summary):
         self.logger.debug(responses_summary)
         return responses_summary
 
-    def purge_url(self, url, servers):
+    def purge_url(self, url, servers, headers=None):
```
Comment (ziollek) [inline vaas-app/src/vaas/purger/purger.py:19]: We assumed that headers contains map <header_name: values_list> for example: 

```
'Accept': ['image/webp', 'image/jpeg']
'Accept-Language': ['pl-PL', 'en-US']
```

So for above input should be prepared purge executions for each combination as below:

### 1.
```
Accept: image/webp
Accept-Language: pl-PL
```

### 2.
```
Accept: image/webp
Accept-Language: en-US
```

### 3.
```
Accept: image/jpeg
Accept-Language: pl-PL
```

### 4.
```
Accept: image/jpeg
Accept-Language: en-US
```

File: vaas-app/src/vaas/purger/purger.py [hunk 2]
```diff
@@ -12,23 +16,34 @@ def log_and_return_data(self, responses_summary):
...
-                    conn.request("PURGE", purge_url, body='', headers=headers)
-                    resp = conn.getresponse().status
-                    data['success'][server.ip] = "varnish http response code: {}, url={}".format(resp, url)
-                except BadStatusLine:
-                    data['error'][server.ip] = "Bad status line from varnish server, url={}".format(url)
-                except Exception as e:
-                    data['error'][server.ip] = "Unexpected error: {}, url={}".format(e, url)
-
+        with PoolExecutor(max_workers=len(servers)) as executor:
+            for _ in executor.map(partial(self.purge_server, url, parsed_url, headers, data), servers):
```
Comment (ziollek) [inline vaas-app/src/vaas/purger/purger.py:27]: Why do You iterate over results if there are irrelevant ?

File: vaas-app/src/vaas/purger/purger.py [hunk 3]
```diff
@@ -12,23 +16,34 @@ def log_and_return_data(self, responses_summary):
...
+                resp, url, sorted(headers.items(), key=lambda x: x[0]), server.ip, server.http_port
+            )
+        except BadStatusLine:
+            data['error'][server.ip] = "Bad status line from varnish server, url={}, headers={}, server={}:{}".format(
+                url, sorted(headers.items(), key=lambda x: x[0]), server.ip, server.http_port
+            )
+        except Exception as e:
+            data['error'][server.ip] = "Unexpected error: {}, url={}, headers={}, server={}:{}".format(
+                e, url, sorted(headers.items(), key=lambda x: x[0]), server.ip, server.http_port
+            )
```
Comment (ziollek) [inline vaas-app/src/vaas/purger/purger.py:49]: Safer way for collect results is returning for parallel function and compose them in main tread (for loop above), especially when you introduce purging of all combinations of headers mentioned above.

File: vaas-app/src/vaas/purger/purger.py (hunk 1 above)
Comment (jtaczanowski) [reply vaas-app/src/vaas/purger/purger.py:20]: Added this functionality in next commits.

File: vaas-app/src/vaas/purger/purger.py (hunk 2 above)
Comment (jtaczanowski) [reply vaas-app/src/vaas/purger/purger.py]: Yes, it's not needed.
I changed this part of code in next commit.

**Pull request merged**